Add `notebook.workingDirectory` setting

austin3dickey opened the pull request:
<!-- Thank you for submitting a pull request.
If this is your first pull request you can find information about
contributing here:
  * https://github.com/posit-dev/positron/blob/main/CONTRIBUTING.md

We recommend synchronizing your branch with the latest changes in the
main branch by either pulling or rebasing.
-->

<!--
  Describe briefly what problem this pull request resolves, or what
  new feature it introduces. Include screenshots of any new or altered
  UI. Link to any GitHub issues but avoid "magic" keywords that will
  automatically close the issue. If there are any details about your
  approach that are unintuitive or you want to draw attention to, please
  describe them here.
-->
Addresses #7988. Adds a `notebook.workingDirectory` setting and hides the `jupyter.notebookFileRoot` setting, which doesn't work in Positron.

### Release Notes

<!--
  Optionally, replace `N/A` with text to be included in the next release notes.
  The `N/A` bullets are ignored. If you refer to one or more Positron issues,
  these issues are used to collect information about the feature or bugfix, such
  as the relevant language pack as determined by Github labels of type `lang: `.
  The note will automatically be tagged with the language.

  These notes are typically filled by the Positron team. If you are an external
  contributor, you may ignore this section.
-->

#### New Features

- Added `notebook.workingDirectory` setting, which can be used to set the default working directory for notebooks (#7988)

#### Bug Fixes

- N/A


### QA Notes

<!--
  Positron team members: please add relevant e2e test tags, so the tests can be
  run when you open this pull request.

  - Instructions: https://github.com/posit-dev/positron/blob/main/test/e2e/README.md#pull-requests-and-test-tags
  - Available tags: https://github.com/posit-dev/positron/blob/main/test/e2e/infra/test-runner/test-tags.ts
-->
@:notebooks @:vscode-settings @:console @:extensions @:interpreter @:win

I added some unit tests and e2e tests!

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->
Some things to try:
- The `jupyter.notebookFileRoot` setting should no longer exist
- Open a workspace, and a new notebook in a subdirectory of that workspace, with a cell containing `%pwd`, and see its value for different values of the new setting (note you need to close/open the notebook for the change to take effect):
  - (unset)
  - `${workspaceFolder}`
  - `${fileDirname}`
  - `${userHome}`
  - `/some/path/that/exists`
  - `/some/path/that/doesnt/exist`
  - whatever else you want to try (variables are defined [here](https://github.com/posit-dev/positron/blob/adcc385b17e3246c5439ea23c01eaf551b57403d/src/vs/workbench/services/configurationResolver/common/configurationResolver.ts#L77))
- The Console and Terminal working directories shouldn't change
- Filename completions should correspond to the correct working directory
- This should work for R (`getwd()`) or Python (`os.getcwd()`) notebooks, and for VSC or Positron notebooks

austin3dickey committed:
add notebook.workingDirectory setting

nstrayer commented on src/vs/workbench/contrib/notebook/browser/notebook.contribution.ts:
```diff
@@ -1311,6 +1315,14 @@ configurationRegistry.registerConfiguration({
 			type: 'string',
 			default: '',
 			tags: ['notebookLayout']
+		},
+		// --- Start Positron ---
+		[NotebookSetting.workingDirectory]: {
+			markdownDescription: nls.localize('notebook.workingDirectory', "Default working directory for notebook kernels. Supports variables like `${workspaceFolder}`. When empty, uses the notebook file's directory. Any change to this setting will apply to future opened notebooks."),
+			type: 'string',
+			default: '',
+			scope: ConfigurationScope.RESOURCE
```
My only thought is it may get confusion when we're all-in on positron notebooks that this setting is in the original vscode notebooks contrib. I'm not sure we will respect _every_ one of the original vscode notebook config options. I _don't_ think it should go in the positron notebooks one though... So this is probably good for now and then maybe in the future we will split out to a new Switzerland positron(the editor)-specific-notebook-configs section?

austin3dickey replied:
yeah, makes sense! We probably will still at least want to use the `notebooks.` prefix right? (I put the setting here because I thought it should start with that prefix since it only applies to notebooks)

nstrayer commented on src/vs/workbench/contrib/notebook/browser/notebook.contribution.ts:
```diff
@@ -1311,6 +1315,14 @@ configurationRegistry.registerConfiguration({
+		},
+		// --- Start Positron ---
+		[NotebookSetting.workingDirectory]: {
+			markdownDescription: nls.localize('notebook.workingDirectory', "Default working directory for notebook kernels. Supports variables like `${workspaceFolder}`. When empty, uses the notebook file's directory. Any change to this setting will apply to future opened notebooks."),
```
Could we link to all the options here? Or just list them all?
Am I correct in thinking these are the relevant ones?
`workspaceFolder, fileDirname, userHome, env:HOME, env:USERPROFILE`.

nstrayer reviewed (dismissed):
This is some good stuff. The tests are great and you even remembered to update the dates in the headers! (2025/2025 points). 

Some comments on the config setting itself.

austin3dickey committed:
link to variable docs

seeM commented on extensions/positron-python/python_files/posit/positron/positron_jedilsp.py:
```diff
@@ -285,10 +286,16 @@ def lsp_initialize(self, params: InitializeParams) -> InitializeResult:
 
-        # If a notebook path was provided, set the project path to the notebook's parent.
-        # See https://github.com/posit-dev/positron/issues/5948.
+        # Set the project path to the notebook's parent if `working_directory` is not provided.
```
What do you think about removing `notebook_path` and handling that case via `working_directory` as well?

seeM reviewed (dismissed):
Nicely done! Works as expected when changing the setting to `${workspaceFolder}` as well as `${workspaceFolder}/mypkg` where `mypkg` exists, in both R and Python.

When I set it to a nonexistent folder like `${workspaceFolder}/mypkg2` or a bad template like `${workspaceFolder2}` it seemed to default to `${workspaceFolder}`, although I don't see where that's handled. Should we handle and/or test those cases?

austin3dickey commented:
>When I set it to a nonexistent folder like `${workspaceFolder}/mypkg2` or a bad template like `${workspaceFolder2}` it seemed to default to `${workspaceFolder}`, although I don't see where that's handled. Should we handle and/or test those cases?

Oh nice catch, I thought I tried this but that's different behavior from what I'd expect (I think I'd expect it to fall back to the parent dir). Turns out this is happening in [kallichore](https://github.com/posit-dev/kallichore/blob/b4fa6e47ea98046394de2228ae4fa252cd9bf788/crates/kcserver/src/kernel_session.rs#L571). If the directory doesn't exist, it defaults to the pwd, which is the workspace root.

Maybe I'll verify that the directory exists at the end of `resolveWorkingDirectory()` so that if it doesn't we can fall back to the old behavior, the parent dir.

austin3dickey committed:
moved the default up the stack

austin3dickey commented:
Okay, I just moved the default behavior of "if the setting doesn't resolve to an existing directory, use the notebook's parent" up the stack to when we're creating the RuntimeSession metadata (in `resolveNotebookWorkingDirectory()`). That way we can delete the code that handles that default in KallichoreSession and the Jedi implementation. This fixes the bug @seeM was seeing.

austin3dickey commented:
I'll have to fix the Python tests and add an e2e test.

austin3dickey committed:
fix python ci

austin3dickey committed:
e2e tests

austin3dickey committed:
fix e2e for windows, maybe

austin3dickey committed:
fix node unit tests

seeM reviewed (dismissed):
Code changes LGTM! Just need to fix the remaining e2e tests, and we might want @rodrigosf672 or @testlabauto to review the new e2e tests

rodrigosf672 commented:
@austin3dickey Thank you so much for adding the e2e tests. For the e2e test `notebook-working-directory.test.ts`, please see the following below. I'll check the other test's failure and investigate the reason as well.

What you're expecting to be visible seems to be inside 5-6 layers of iframes/shadow DOMs. Essentially, it's visible on the trace, but not for Playwright. I'd try piercing through the iframes/shadow DOMs by using the following locator.
`locator('.cell-editor-container > .monaco-editor > .overflow-guard > div:nth-child(3) > .margin-view-overlays > div > .current-line')`

Alternatively, you may use text that is present in what you are trying to locate (_e.g.,_ `%pwd`). The reason why text could work well is because, since Playwright 1.26, text locators are automatically pierced through ([1](https://playwright.dev/docs/release-notes#version-126),[2](https://playwright.dev/docs/api/class-locatorassertions#locator-assertions-to-have-text)). 

These are just thoughts I have after a quick glance; I'd need to check them further to be sure. Do you want me to fix it or do you want to give it a try? Please let me know and thank you so much for adding the tests in the first place! 👍

austin3dickey committed:
fix e2e

austin3dickey commented:
Hey @rodrigosf672 thanks for the advice! In my case it's because the expected regex wasn't matching the output for those 2 failing tests, but it was for the other 4. I just updated the expected output so they all pass locally. 

I also deleted the `userHome` test; it feels like overkill because we have so many other working tests and we'd need to handle the behavior differently across OSes.

austin3dickey commented:
Okay, now I *think* CI will finally be green. 🙂 However, after all this, I think I'll merge this later this week after we cut the 2025.08 release branch, so that this change ships with 09. It touches just a little too much code for me to be 100% confident that 08 is a good fit at this point. Thanks @seeM for the suggestion.

austin3dickey committed:
forgot a copyright header
